{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b35cc53d",
   "metadata": {},
   "source": [
    "### Initialize a list of queries (size = num examples available)\n",
    "\n",
    "##### These will be parsed to generate core queries + ways to augment then"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1b9f0d7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bgee queries (with variants) from tutorial online: \n",
    "# https://www.bgee.org/support/tutorial-query-bgee-knowledge-graph-sparql#querying-with-controlled-vocabularies-and-identifiers\n",
    "\n",
    "qs = [None] * 16 \n",
    "\n",
    "qs[0] = []\n",
    "\n",
    "qs[1] = [\"\"\"\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "SELECT ?species {\n",
    "\t?species a up:Taxon .}\n",
    "    \"\"\"]\n",
    "\n",
    "qs[2] = [\"\"\"\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "SELECT ?species ?sci_name ?common_name {\n",
    "\t?species a up:Taxon .\n",
    "\t?species up:scientificName ?sci_name .\n",
    "\t?species up:rank up:Species .\n",
    "\tOPTIONAL { ?species up:commonName ?common_name .}\n",
    "}\n",
    "\"\"\"]\n",
    "\n",
    "qs[3] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "SELECT DISTINCT ?anat ?anatName {\n",
    "\t?seq a orth:Gene .\n",
    "\t?seq genex:isExpressedIn ?anat .\n",
    "\t?seq rdfs:label \"APOC1\" .\n",
    "\t?anat a genex:AnatomicalEntity .\n",
    "\t?anat rdfs:label ?anatName .\n",
    "}\n",
    "\n",
    "\"\"\"]\n",
    "\n",
    "qs[4] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName {\n",
    "\t?seq a orth:Gene .\n",
    "\t?seq genex:isExpressedIn ?anat.\n",
    "\t?seq rdfs:label \"APOC1\" .\n",
    "\t?anat a genex:AnatomicalEntity .\n",
    "\t?anat rdfs:label ?anatName .\n",
    "\t\t### Specifying species:\n",
    "\t\t?seq orth:organism ?organism .\n",
    "\t\t?organism obo:RO_0002162  ?species .\n",
    "\t\t?species a up:Taxon .\n",
    "\t\t?species up:scientificName \"Homo sapiens\" .\n",
    "}\n",
    "\"\"\"]\n",
    "\n",
    "qs[5] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX efo: <http://www.ebi.ac.uk/efo/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName {\n",
    "\t?seq a orth:Gene .\n",
    "\t?seq genex:isExpressedIn ?condition.\n",
    "    ?condition a genex:ExpressionCondition.\n",
    "\t?seq rdfs:label \"APOC1\" .\n",
    "\t?condition genex:hasAnatomicalEntity ?anat .\n",
    "    ?anat a genex:AnatomicalEntity.\n",
    "\t?condition genex:hasAnatomicalEntity obo:GO_0005575 .\n",
    "\t?anat rdfs:label ?anatName .\n",
    "\t?condition genex:hasDevelopmentalStage ?stage .\n",
    "    ?stage a efo:EFO_0000399 .\n",
    "\t?stage rdfs:label \"life cycle\" .\n",
    "\t?condition genex:hasSex \"any\".\n",
    "\t?condition genex:hasStrain ?strain .\n",
    "\t?strain rdfs:label \"wild-type\" .\n",
    "}\n",
    "\"\"\"]\n",
    "\n",
    "qs[6] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX efo: <http://www.ebi.ac.uk/efo/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName ?stage {\n",
    "\t?seq a orth:Gene .\n",
    "\t?seq genex:isExpressedIn ?condition.\n",
    "    ?condition a genex:ExpressionCondition.\n",
    "\t?seq rdfs:label \"APOC1\" .\n",
    "\t?condition genex:hasAnatomicalEntity ?anat .\n",
    "\t?anat rdfs:label ?anatName .\n",
    "\t\t?condition genex:hasAnatomicalEntity obo:GO_0005575 .\n",
    "\t\t?condition genex:hasDevelopmentalStage ?stage .\n",
    "        ?stage a efo:EFO_0000399 .\n",
    "\t?stage rdfs:label \"post-juvenile\" .\n",
    "\t?condition genex:hasSex \"any\".\n",
    "\t?condition genex:hasStrain ?strain .\n",
    "    ?strain a efo:EFO_0005135.\n",
    "\t\t?strain rdfs:label \"wild-type\" .\n",
    "\t\t?seq orth:organism ?organism .\n",
    "\t\t?organism obo:RO_0002162  ?species .\n",
    "\t\t?species a up:Taxon .\n",
    "\t\t?species up:commonName \"human\" .\n",
    "}\n",
    "\"\"\"]\n",
    "\n",
    "qs[7] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX efo: <http://www.ebi.ac.uk/efo/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName ?stage {\n",
    "\t?seq a orth:Gene .\n",
    "\t?seq genex:isExpressedIn ?condition.\n",
    "\t?seq rdfs:label \"APOC1\" .\n",
    "    ?condition a genex:ExpressionCondition.\n",
    "\t?condition genex:hasAnatomicalEntity ?anat .\n",
    "    ?anat a genex:AnatomicalEntity.\n",
    "\t?anat rdfs:label ?anatName .\n",
    "\t\t?condition genex:hasAnatomicalEntity obo:GO_0005575 .\n",
    "\t\t?condition genex:hasDevelopmentalStage ?stage .\n",
    "        ?stage a efo:EFO_0000399 .\n",
    "\t?stage rdfs:label \"post-juvenile\" .\n",
    "\t\t?seq orth:organism ?organism .\n",
    "\t\t?organism obo:RO_0002162  ?species .\n",
    "\t\t?species a up:Taxon .\n",
    "\t\t?species up:commonName \"human\" .\n",
    "}\n",
    "\"\"\"]\n",
    "\n",
    "qs[8] = [None] * 3\n",
    "\n",
    "qs[8][0] = \"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName ?score ?stage {\n",
    "\t?seq a orth:Gene .\n",
    "\t?expression a genex:Expression .\n",
    "\t?expression genex:hasExpressionCondition ?condition .\n",
    "\t?expression genex:hasExpressionLevel ?score .\n",
    "\t?expression genex:hasSequenceUnit ?seq .\n",
    "\t?seq rdfs:label \"APOC1\" .\n",
    "\t?condition a genex:ExpressionCondition.\n",
    "\t?condition genex:hasAnatomicalEntity ?anat .\n",
    "\t?condition genex:hasAnatomicalEntity obo:GO_0005575 .\n",
    "\t?anat a genex:AnatomicalEntity.\n",
    "\t?anat rdfs:label ?anatName .\n",
    "\t?condition genex:hasDevelopmentalStage ?stage .\n",
    "\t?stage rdfs:label \"post-juvenile\" .\n",
    "\t?condition genex:hasSex \"any\".\n",
    "\t?condition genex:hasStrain ?strain .\n",
    "\t?strain rdfs:label \"wild-type\" .\n",
    "\t?seq orth:organism ?organism .\n",
    "\t?organism obo:RO_0002162  ?species .\n",
    "\t?species a up:Taxon .\n",
    "\t?species up:commonName \"human\" .\n",
    "FILTER (?anat !=  obo:GO_0005575)\n",
    "} ORDER BY DESC(?score)\n",
    "\"\"\"\n",
    "\n",
    "qs[8][1] = \"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX efo: <http://www.ebi.ac.uk/efo/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName ?score  {\n",
    "\t?seq a orth:Gene .\n",
    "\t?expression a genex:Expression .\n",
    "\t?expression genex:hasExpressionCondition ?condition .\n",
    "\t?expression genex:hasExpressionLevel ?score .\n",
    "\t?expression genex:hasSequenceUnit ?seq .\n",
    "\t?seq lscr:xrefEnsemblGene ensembl:ENSG00000130208 .\n",
    "\t?condition a genex:ExpressionCondition.\n",
    "\t?condition genex:hasAnatomicalEntity ?anat .\n",
    "\t?anat a genex:AnatomicalEntity.\n",
    "\t?anat rdfs:label ?anatName .\n",
    "\t?condition genex:hasDevelopmentalStage obo:UBERON_0000113 .\n",
    "\t?condition genex:hasSex \"any\".\n",
    "\t?condition genex:hasStrain ?strain .\n",
    "    ?strain a efo:EFO_0005135.\n",
    "\t?strain rdfs:label \"wild-type\" .\n",
    "\t?seq orth:organism ?organism .\n",
    "\t?organism obo:RO_0002162  up-taxon:9606 .\n",
    "    FILTER (?anat !=  obo:GO_0005575)\n",
    "} ORDER BY DESC(?score)\n",
    "\"\"\"\n",
    "\n",
    "qs[8][2] = \"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName ?score  {\n",
    "\t?seq a orth:Gene .\n",
    "\t?expression a genex:Expression .\n",
    "\t?expression genex:hasExpressionCondition ?condition .\n",
    "\t?expression genex:hasExpressionLevel ?score .\n",
    "\t?expression genex:hasSequenceUnit ?seq .\n",
    "\t?seq lscr:xrefEnsemblGene ensembl:ENSG00000130208 .\n",
    "    ?condition a genex:ExpressionCondition.\n",
    "\t?condition genex:hasAnatomicalEntity ?anat .\n",
    "    ?anat a genex:AnatomicalEntity.\n",
    "\t?anat rdfs:label ?anatName .\n",
    "\t?condition genex:hasDevelopmentalStage obo:UBERON_0000113 .\n",
    "\t?condition genex:hasSex \"any\".\n",
    "\t?condition genex:hasStrain ?strain .\n",
    "    ?strain a efo:EFO_0005135.\n",
    "\t?strain rdfs:label \"wild-type\" .\n",
    "FILTER (?anat !=  obo:GO_0005575)\n",
    "} ORDER BY DESC(?score)\n",
    "\"\"\"\n",
    "\n",
    "qs[9] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?cellType ?anatName ?cellTypeName ?score ?stage {\n",
    "\t?seq a orth:Gene .\n",
    "\t?expression a genex:Expression .\n",
    "\t?expression genex:hasExpressionCondition ?condition .\n",
    "\t?expression genex:hasExpressionLevel ?score .\n",
    "\t?expression genex:hasSequenceUnit ?seq .\n",
    "\t?seq rdfs:label \"APOC1\" .\n",
    "    ?condition a genex:ExpressionCondition.\n",
    "\t?condition genex:hasAnatomicalEntity ?anat .\n",
    "    ?anat a genex:AnatomicalEntity.\n",
    "\t?anat rdfs:label ?anatName .\n",
    "\t?condition genex:hasAnatomicalEntity ?cellType .\n",
    "\t?cellType rdfs:label ?cellTypeName .\n",
    "\t?condition genex:hasDevelopmentalStage ?stage .\n",
    "\t?stage rdfs:label \"post-juvenile\" .\n",
    "\t?condition genex:hasSex \"any\".\n",
    "\t?condition genex:hasStrain ?strain .\n",
    "    ?strain a efo:EFO_0005135.\n",
    "\t?strain rdfs:label \"wild-type\" .\n",
    "\t?seq orth:organism ?organism .\n",
    "\t?organism obo:RO_0002162  ?species .\n",
    "\t?species a up:Taxon .\n",
    "\t?species up:commonName \"human\" .\n",
    "    FILTER (?anat != obo:GO_0005575)\n",
    "    FILTER (?anat != ?cellType)\n",
    "} ORDER BY DESC(?score)\n",
    "\"\"\"]\n",
    "\n",
    "qs[10] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?cellType ?anatName ?cellTypeName ?score ?stage {\n",
    "\t?seq a orth:Gene .\n",
    "\t?expression a genex:Expression .\n",
    "\t?expression genex:hasExpressionCondition ?condition .\n",
    "\t?expression genex:hasExpressionLevel ?score .\n",
    "\t?expression genex:hasSequenceUnit ?seq .\n",
    "\t?seq rdfs:label \"APOC1\" .\n",
    "    ?condition a genex:ExpressionCondition.\n",
    "\t?condition genex:hasAnatomicalEntity ?anat .\n",
    "    ?anat a genex:AnatomicalEntity.\n",
    "\t?anat rdfs:label ?anatName .\n",
    "\t?condition genex:hasAnatomicalEntity ?cellType .\n",
    "\t?cellType rdfs:label ?cellTypeName .\n",
    "\t?condition genex:hasDevelopmentalStage ?stage .\n",
    "    ?stage rdf:type efo:EFO_0000399 .\n",
    "\t?stage rdfs:label \"post-juvenile\" .\n",
    "\t?seq orth:organism ?organism .\n",
    "\t?organism obo:RO_0002162  ?species .\n",
    "\t?species a up:Taxon .\n",
    "\t?species up:commonName \"human\" .\n",
    "    FILTER (?anat !=  obo:GO_0005575)\n",
    "    FILTER (?anat != ?cellType)\n",
    "} ORDER BY DESC(?score)\n",
    "\"\"\"]\n",
    "\n",
    "qs[11] = [None] * 2\n",
    "\n",
    "qs[11][0] = \"\"\"\n",
    "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "PREFIX efo: <http://www.ebi.ac.uk/efo/EFO_0000399>\n",
    "\n",
    "SELECT DISTINCT ?stage ?stageName ?stageDescription {\n",
    "        ?stage rdf:type efo:EFO_0000399 .\n",
    "\t\t?stage rdfs:label ?stageName .\n",
    "\t\t?stage dcterms:description ?stageDescription .\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "qs[11][1] = \"\"\"\n",
    "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "PREFIX efo: <http://www.ebi.ac.uk/efo/EFO_0000399>\n",
    "\n",
    "SELECT DISTINCT ?stage ?stageName ?stageDescription {\n",
    "        ?stage rdf:type efo:EFO_0000399 . \n",
    "\t\t?stage rdfs:label ?stageName .\n",
    "\t\t?stage dcterms:description ?stageDescription .\n",
    "\t\tFILTER (CONTAINS(?stageName,\"post-juvenile\"))\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "qs[12] = [None] * 3\n",
    "\n",
    "qs[12][0] = \"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName ?stageIRI ?score  {\n",
    "\t?seq a orth:Gene .\n",
    "\t?expression a genex:Expression .\n",
    "\t?expression genex:hasExpressionCondition ?condition .\n",
    "\t?expression genex:hasExpressionLevel ?score .\n",
    "\t?expression genex:hasSequenceUnit ?seq .\n",
    "\t?seq lscr:xrefNCBIGene <https://www.ncbi.nlm.nih.gov/gene/118230125> .\n",
    "\t?condition genex:hasAnatomicalEntity ?anat .\n",
    "    ?anat a genex:AnatomicalEntity.\n",
    "\t?anat rdfs:label ?anatName .\n",
    "    ?condition a genex:ExpressionCondition.\n",
    "\t?condition genex:hasDevelopmentalStage ?stageIRI .\n",
    "    ?stageIRI rdf:type efo:EFO_0000399 . \n",
    "\t?condition genex:hasSex \"any\".\n",
    "\t?condition genex:hasStrain ?strain .\n",
    "    ?strain a efo:EFO_0005135.\n",
    "\t?strain rdfs:label \"wild-type\" .\n",
    "    FILTER (?anat != obo:GO_0005575)\n",
    "} ORDER BY DESC(?score)\n",
    "\"\"\"\n",
    "\n",
    "qs[12][1] = \"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName ?stageIRI ?score  {\n",
    "\t?seq a orth:Gene .\n",
    "\t?expression a genex:Expression .\n",
    "\t?expression genex:hasExpressionCondition ?condition .\n",
    "\t?expression genex:hasExpressionLevel ?score .\n",
    "\t?expression genex:hasSequenceUnit ?seq .\n",
    "\t?seq dcterms:identifier \"118230125\" .\n",
    "    ?condition a genex:ExpressionCondition.\n",
    "\t?condition genex:hasAnatomicalEntity ?anat .\n",
    "\t?anat rdfs:label ?anatName .\n",
    "\t?condition genex:hasDevelopmentalStage ?stageIRI .\n",
    "\t?condition genex:hasSex \"any\".\n",
    "\t?condition genex:hasStrain ?strain .\n",
    "    ?strain a efo:EFO_0005135.\n",
    "\t?strain rdfs:label \"wild-type\" .\n",
    "FILTER (?anat !=  obo:GO_0005575)\n",
    "} ORDER BY DESC(?score)\n",
    "\"\"\"\n",
    "\n",
    "qs[13] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up-protein:<http://purl.uniprot.org/uniprot/>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName {\n",
    "\t?seq a orth:Gene .\n",
    "\t?seq genex:isExpressedIn ?anat .\n",
    "\t?seq lscr:xrefUniprot up-protein:P02654 .\n",
    "\t?anat a genex:AnatomicalEntity .\n",
    "\t?anat rdfs:label ?anatName .\n",
    "}\n",
    "\"\"\"]\n",
    "\n",
    "qs[14] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "\n",
    "SELECT DISTINCT ?symbol ?description ?id\n",
    "?links ?organism ?uniprot ?ensembl ?ncbi  {\n",
    "    ?seq a orth:Gene .\n",
    "    ?seq rdfs:label ?symbol .\n",
    "    ?seq rdfs:seeAlso ?links .\n",
    "    ?seq dcterms:description ?description .\n",
    "    ?seq dcterms:identifier ?id .\n",
    "    ?seq orth:organism ?organism .\n",
    "    OPTIONAL{?seq lscr:xrefUniprot ?uniprot .}\n",
    "    OPTIONAL{?seq lscr:xrefEnsemblGene ?ensembl .}\n",
    "    OPTIONAL{?seq lscr:xrefNCBIGene ?ncbi .}\n",
    "    FILTER (?id = \"ENSG00000130208\")\n",
    "}\n",
    "\"\"\"]\n",
    "\n",
    "qs[15] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName {\n",
    "\t?seq a orth:Gene .\n",
    "\t?seq genex:isAbsentIn ?anat.\n",
    "\t?seq rdfs:label \"APOC1\" .\n",
    "\t?anat a genex:AnatomicalEntity .\n",
    "\t?anat rdfs:label ?anatName .\n",
    "\t?seq orth:organism ?organism .\n",
    "\t?organism obo:RO_0002162  ?species .\n",
    "\t?species a up:Taxon .\n",
    "\t?species up:scientificName \"Homo sapiens\" .\n",
    "}\n",
    "\"\"\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "eed8c0b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# also store NL questions corresponding to the queries\n",
    "\n",
    "nlqs = [None] *16\n",
    "\n",
    "nlqs[0] = \"\"\n",
    "nlqs[1] = \"What are the species present in Bgee?\"\n",
    "nlqs[2] = \"What are the species present in Bgee and their scientific and common names?\"\n",
    "nlqs[3] = \"What are the anatomical entities where the APOC1 gene is expressed?\"\n",
    "nlqs[4] = \"What are the anatomical entities where the APOC1 Homo sapiens gene is expressed?\"\n",
    "nlqs[5] = \"What are the anatomical entities where the APOC1 gene is expressed independently of the developmental stage, sex, strain and cell type?\"\n",
    "nlqs[6] = \"What are the anatomical entities where the human gene APOC1 is expressed in the post-juvenile stage?\"\n",
    "nlqs[7] = \"What are the anatomical entities where the human gene APOC1 is expressed in the post-juvenile stage?\"\n",
    "nlqs[8] = \"What are the anatomical entities where the human gene APOC1 is expressed in the post-juvenile stage along with its expression score independently of the strain, sex, and cell type?\"\n",
    "nlqs[9] = \"What are the anatomical entities including cell types, if any, where the human gene APOC1 is expressed at the post-juvenile stage along with its expression score independently of the strain and sex?\"\n",
    "nlqs[10] = \"What are the anatomical entities including cell types, if any, where the human gene APOC1 is expressed at the post-juvenile stage along with its expression score independently of the strain and sex?\"\n",
    "nlqs[11] = \"What are the developmental stages present in Bgee?\"\n",
    "nlqs[12] = \"What are the anatomical entities where the eel gene apoc1 is expressed along with its expression score independently of the strain, sex, and cell type?\"\n",
    "nlqs[13] = \"What are the anatomical entities where the P02654 gene is expressed? Note that P02654 is a UniProtKB identifier of the APOC1 human gene.\"\n",
    "nlqs[14] = \"What is all the metadata related to the ENSG00000130208 gene, where ENSG00000130208 is the identifier of the APOC1 human gene. \"\n",
    "nlqs[15] = \"What are the anatomical entities where the APOC1 Homo sapiens gene is not expressed, that is where is APOC1 absent ?\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5aa0300c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. construct augmented SPARQL query\n",
    "\n",
    "def construct_query(prefixes, vars_list, triples, optionals, filters):\n",
    "    sparql_query = ''\n",
    "    \n",
    "    for prefix in prefixes :\n",
    "        sparql_query += prefix + \"\\n\"\n",
    "    \n",
    "    select_stmt = \"SELECT DISTINCT \"\n",
    "\n",
    "    for var in vars_list:\n",
    "        select_stmt += \"?\" + var + \" \"\n",
    "            \n",
    "    select_stmt += \"WHERE \"\n",
    "    \n",
    "    sparql_query += select_stmt + \"{ \\n\"\n",
    "\n",
    "    \n",
    "    for triple in triples:\n",
    "        sparql_query += \"\\t\" + triple + \"\\n\"\n",
    "    \n",
    "    for optional in optionals:\n",
    "        sparql_query += \"\\t\" + \"OPTIONAL { \" + optional + \" } \" + \"\\n\"\n",
    "        # here add also class...property label (if it exists)...or COMMENT for the property label e.g. #encoded by\n",
    "     \n",
    "    for filter in filters:\n",
    "        sparql_query += \"\\t\" + \"FILTER( \" + filter + \" )\" + \"\\n\"\n",
    "        \n",
    "    sparql_query += \"}\"\n",
    "    \n",
    "    return sparql_query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8aa338ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: SPARQLWrapper in /Users/asima/opt/anaconda3/lib/python3.9/site-packages (1.8.5)\n",
      "Requirement already satisfied: rdflib>=4.0 in /Users/asima/opt/anaconda3/lib/python3.9/site-packages (from SPARQLWrapper) (6.2.0)\n",
      "Requirement already satisfied: pyparsing in /Users/asima/opt/anaconda3/lib/python3.9/site-packages (from rdflib>=4.0->SPARQLWrapper) (3.0.4)\n",
      "Requirement already satisfied: setuptools in /Users/asima/opt/anaconda3/lib/python3.9/site-packages (from rdflib>=4.0->SPARQLWrapper) (58.0.4)\n",
      "Requirement already satisfied: isodate in /Users/asima/opt/anaconda3/lib/python3.9/site-packages (from rdflib>=4.0->SPARQLWrapper) (0.6.1)\n",
      "Requirement already satisfied: six in /Users/asima/opt/anaconda3/lib/python3.9/site-packages (from isodate->rdflib>=4.0->SPARQLWrapper) (1.16.0)\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Value must be a nonnegative integer or None",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/5d/3j6fyxq90gs2qdxgzpzsl1gh0000gp/T/ipykernel_45025/2228074716.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0mNUM_EXAMPLES\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m10\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;31m# always display full column results (don't truncate output)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 10\u001b[0;31m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_option\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'display.max_colwidth'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;31m# the endpoints must be defined as wrappers for executing SPARQL queries\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/_config/config.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwds)\u001b[0m\n\u001b[1;32m    259\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    260\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__call__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mT\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 261\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__func__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    262\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    263\u001b[0m     \u001b[0;31m# error: Signature of \"__doc__\" incompatible with supertype \"object\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/_config/config.py\u001b[0m in \u001b[0;36m_set_option\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    158\u001b[0m         \u001b[0mo\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_get_registered_option\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    159\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mo\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mo\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalidator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 160\u001b[0;31m             \u001b[0mo\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalidator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    161\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    162\u001b[0m         \u001b[0;31m# walk the nested dict\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/_config/config.py\u001b[0m in \u001b[0;36mis_nonnegative_int\u001b[0;34m(value)\u001b[0m\n\u001b[1;32m    880\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    881\u001b[0m     \u001b[0mmsg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"Value must be a nonnegative integer or None\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 882\u001b[0;31m     \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmsg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    883\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    884\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Value must be a nonnegative integer or None"
     ]
    }
   ],
   "source": [
    "# first we need to define the SPARQL endpoints of each source, to use later in the protocols\n",
    "import sys\n",
    "!{sys.executable} -m pip install SPARQLWrapper\n",
    "from SPARQLWrapper import SPARQLWrapper, JSON, CSV\n",
    "import sys, os, time\n",
    "import pandas as pd\n",
    "\n",
    "NUM_EXAMPLES = 10\n",
    "# always display full column results (don't truncate output)\n",
    "pd.set_option('display.max_colwidth', -1)\n",
    "\n",
    "# the endpoints must be defined as wrappers for executing SPARQL queries\n",
    "sparql_OMA = SPARQLWrapper(\"https://sparql.omabrowser.org/sparql\")\n",
    "sparql_Bgee = SPARQLWrapper(\"https://www.bgee.org/sparql/\")\n",
    "sparql_RIKEN = SPARQLWrapper(\"https://knowledge.brc.riken.jp/sparql\")\n",
    "\n",
    "# function to print in a table results of a SPARQL query\n",
    "def pretty_print(results):\n",
    "    \n",
    "    # how to transform SPARQL results into Pandas dataframes\n",
    "    \n",
    "    # get header (column names) from results\n",
    "    header = results[\"results\"][\"bindings\"][0].keys()\n",
    "\n",
    "    # display table of results:\n",
    "    table = []\n",
    "    \n",
    "    # the SPARQL JSON results to the query are available in the \"results\", \"bindings\" entry:\n",
    "    for entry in results[\"results\"][\"bindings\"]:\n",
    "        # append entries from the results to a regular Python list of rows, which we can then transform to a Pandas DF\n",
    "        row = [entry[column][\"value\"] if entry.get(column, None) != None else None for column in header]\n",
    "        table.append(row)\n",
    "    df = pd.DataFrame(table, columns=list(header))\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5b21b139",
   "metadata": {},
   "outputs": [],
   "source": [
    "def execute_query(sparql_endpoint, query):\n",
    "    sparql_endpoint.setQuery(query)\n",
    "    sparql_endpoint.setReturnFormat(JSON)\n",
    "\n",
    "    results = sparql_endpoint.query().convert()\n",
    "    return results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f01f16f",
   "metadata": {},
   "source": [
    "### TODOs\n",
    "- identify parts in graph where the literal properties occur - assign comments with the appropriate labels to them\n",
    "=> augmented set 1\n",
    "- break down queries into more general ones (fewer triples)\n",
    "- progressively add more triple patterns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61fab8fd",
   "metadata": {},
   "source": [
    "How to? example: https://pypi.org/project/SPARQL-parser/ . Library doesn't work but we reuse code from it here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "60253e72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['PREFIX orth: <http://purl.org/net/orth#>',\n",
       "  'PREFIX genex: <http://purl.org/genex#>',\n",
       "  'PREFIX obo: <http://purl.obolibrary.org/obo/>',\n",
       "  'PREFIX up: <http://purl.uniprot.org/core/>',\n",
       "  'PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>',\n",
       "  'PREFIX lscr: <http://purl.org/lscr#>',\n",
       "  'PREFIX dcterms: <http://purl.org/dc/terms/>'],\n",
       " {'anat',\n",
       "  'anatName',\n",
       "  'condition',\n",
       "  'expression',\n",
       "  'score',\n",
       "  'seq',\n",
       "  'stageIRI',\n",
       "  'strain'},\n",
       " ['?seq a orth:Gene .',\n",
       "  '?expression a genex:Expression .',\n",
       "  '?expression genex:hasExpressionCondition ?condition .',\n",
       "  '?expression genex:hasExpressionLevel ?score .',\n",
       "  '?expression genex:hasSequenceUnit ?seq .',\n",
       "  '?seq dcterms:identifier \"118230125\" .',\n",
       "  '?condition genex:hasAnatomicalEntity ?anat .',\n",
       "  '?anat a genex:AnatomicalEntity.',\n",
       "  '?anat rdfs:label ?anatName .',\n",
       "  '?condition a genex:ExpressionCondition.',\n",
       "  '?condition genex:hasDevelopmentalStage ?stageIRI .',\n",
       "  '?condition genex:hasSex \"any\".',\n",
       "  '?condition genex:hasStrain ?strain .',\n",
       "  '?strain rdfs:label \"wild-type\" .'],\n",
       " [],\n",
       " ['?anat !=  obo:GO_0005575']]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "def get_core_triples(sparql_query):\n",
    "    # separate prefixes\n",
    "    # get triples as list of strings\n",
    "    # NOTE: must parse OPTIONALS separately too\n",
    "    # these should be handled separately, since we want to add them only as optionals\n",
    "    query_string = sparql_query\n",
    "    \n",
    "    prefixes_string = query_string[:query_string.lower().find('select')]\n",
    "    prefixes = prefixes_string.split(\"\\n\")\n",
    "    prefixes = [prefix for prefix in prefixes if prefix != '']\n",
    "    \n",
    "    # get variables mentioned in query\n",
    "    vars_list = set(re.findall(r'\\?([a-z_0-9A-Z]+)', query_string))\n",
    "    \n",
    "    # split query by \". or ;\"\n",
    "    if query_string.find('{') != -1:  \n",
    "        query_start = query_string.find('{') + 1\n",
    "        query_end = query_string.rfind(\"}\")\n",
    "        \n",
    "    query_string_core = query_string[query_start : query_end].replace(\"\\t\", \"\")\n",
    "\n",
    "    # SPLIT triple patterns. assumption: \"nicely written\" queries, with 1 triple pattern per line\n",
    "    triples = query_string_core.split(\"\\n\")\n",
    "    \n",
    "    triples = [triple.strip() for triple in triples if triple and not triple.isspace()]\n",
    "    \n",
    "    #print(\"Query fragments (triple patterns, OPTIONALs, FILTERs): \\n{}\".format(triples))\n",
    "    \n",
    "    optionals = []\n",
    "    filters = []\n",
    "    \n",
    "    final_triples = []\n",
    "    \n",
    "    # also return FILTERs separately\n",
    "    for triple in triples:\n",
    "        triple = triple.strip()\n",
    "        if(triple.startswith(\"}\")):\n",
    "            triple = triple[1:].strip()\n",
    "        if(triple.lower().find(\"filter\") != -1):\n",
    "            matches = re.findall(r'\\((.*?)\\)',triple)\n",
    "            for filter in matches:\n",
    "                filters += [filter]\n",
    "            continue\n",
    "        if(triple.lower().startswith(\"optional\")):\n",
    "            optional_start = triple.find(\"{\") + 1\n",
    "            optionals += [triple[optional_start:triple.rfind(\"}\")].strip()]\n",
    "            continue\n",
    "        final_triples += [triple]\n",
    "\n",
    "    return [prefixes, vars_list, final_triples, optionals, filters]\n",
    "    \n",
    "\n",
    "query = \"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName ?stageIRI ?score  {\n",
    "\t?seq a orth:Gene .\n",
    "\t\t?expression a genex:Expression .\n",
    "\t\t?expression genex:hasExpressionCondition ?condition .\n",
    "\t\t?expression genex:hasExpressionLevel ?score .\n",
    "\t?expression genex:hasSequenceUnit ?seq .\n",
    "\t\t?seq dcterms:identifier \"118230125\" .\n",
    "\t?condition genex:hasAnatomicalEntity ?anat .\n",
    "    ?anat a genex:AnatomicalEntity.\n",
    "\t\t?anat rdfs:label ?anatName .\n",
    "        ?condition a genex:ExpressionCondition.\n",
    "\t\t?condition genex:hasDevelopmentalStage ?stageIRI .\n",
    "\t?condition genex:hasSex \"any\".\n",
    "\t?condition genex:hasStrain ?strain .\n",
    "\t\t?strain rdfs:label \"wild-type\" .\n",
    "FILTER (?anat !=  obo:GO_0005575)\n",
    "} ORDER BY DESC(?score)\n",
    "\"\"\"\n",
    "\n",
    "get_core_triples(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a99e2b64",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_name_from_URI(Uri):\n",
    "    if(not Uri.startswith(\"http\")): # defined via a PREFIX statement\n",
    "        return Uri[Uri.find(\":\") + 1 : ]\n",
    "    fragmentName = urlparse(Uri).fragment\n",
    "    if fragmentName:\n",
    "        return fragmentName\n",
    "    else:\n",
    "        return Uri[Uri.rfind(\"/\") + 1 : ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ac46cb0",
   "metadata": {},
   "source": [
    "Note: Simplifying Assumption: the variable must have a type so that it can be augmented with a description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fdbd7796",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                           class classLabel  \\\n",
      "0  http://purl.org/net/orth#Gene       Gene   \n",
      "1  http://purl.org/net/orth#Gene       Gene   \n",
      "2  http://purl.org/net/orth#Gene       Gene   \n",
      "\n",
      "                                     property propertyLabel  \n",
      "0  http://www.w3.org/2000/01/rdf-schema#label         label  \n",
      "1         http://purl.org/dc/terms/identifier    Identifier  \n",
      "2        http://purl.org/dc/terms/description   Description  \n",
      "{'Gene': {'classLabel': {'gene', 'Gene'}, 'propNamesLabels': {'http://www.w3.org/2000/01/rdf-schema#label': 'label', 'http://purl.org/dc/terms/identifier': 'Identifier', 'http://purl.org/dc/terms/description': 'Description', 'http://www.w3.org/2000/01/rdf-schema#seeAlso': 'see also'}}, 'AnatomicalEntity': {'classLabel': {'Anatomical entity'}, 'propNamesLabels': {'http://purl.org/dc/terms/description': 'Description', 'http://www.w3.org/2000/01/rdf-schema#label': 'label'}}, 'AbsenceExpression': {'classLabel': {'absence of gene expression'}, 'propNamesLabels': {'http://purl.org/genex#hasFDRpvalue': 'has false discovery rate p-value', 'http://purl.org/genex#hasExpressionLevel': 'has expression level score'}}, 'EFO_0000399': {'classLabel': {'developmental stage'}, 'propNamesLabels': {'http://purl.org/dc/terms/description': 'Description', 'http://www.w3.org/2000/01/rdf-schema#label': 'label'}}, 'Taxon': {'classLabel': {'Taxon'}, 'propNamesLabels': {'http://purl.uniprot.org/core/commonName': 'common name', 'http://purl.uniprot.org/core/scientificName': 'scientific name'}}, 'EFO_0005135': {'classLabel': {'strain'}, 'propNamesLabels': {'http://www.w3.org/2000/01/rdf-schema#label': 'label'}}, 'Expression': {'classLabel': {'gene expression'}, 'propNamesLabels': {'http://purl.org/genex#hasFDRpvalue': 'has false discovery rate p-value', 'http://purl.org/genex#hasExpressionLevel': 'has expression level score'}}}\n"
     ]
    }
   ],
   "source": [
    "# How to augment? \n",
    "from urllib.parse import urlparse, urldefrag, parse_qs\n",
    "\n",
    "# parse file with extra text properties to include in the query\n",
    "\n",
    "import pandas as pd\n",
    "df_extra_props = pd.read_csv ('extra_props_bgee.csv')\n",
    "df_extra_props = df_extra_props.drop_duplicates()\n",
    "print(df_extra_props.head(3))\n",
    "classNameURIs = df_extra_props['class'].drop_duplicates().tolist()\n",
    "\n",
    "# build a json from className to label, props_List : {URI, label}\n",
    "classNamesProps = {}\n",
    "\n",
    "for classNameURI in classNameURIs:\n",
    "    className = extract_name_from_URI(classNameURI)\n",
    "    propNames = df_extra_props[df_extra_props['class']== classNameURI]\n",
    "    classNamesProps[className] = {}\n",
    "    classNamesProps[className][\"classLabel\"] = set(propNames[\"classLabel\"].tolist())\n",
    "    classNamesProps[className][\"propNamesLabels\"] = {x[\"property\"] : x[\"propertyLabel\"] for index, x in propNames.iterrows()}\n",
    "    \n",
    "print(classNamesProps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "91d49426",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "SELECT DISTINCT ?anatomicalentity ?anatomicalentityName ?efo_343_4343 ?anatomicalentity_label ?gene ?anatomicalentity_description ?gene_seealso WHERE { \n",
      "\t?gene a orth:Gene .\n",
      "\t?gene genex:isExpressedIn ?anatomicalentity .\n",
      "\t?gene rdfs:label \"APOC1\" .\n",
      "\t?anatomicalentity a genex:AnatomicalEntity .\n",
      "\t?anatomicalentity rdfs:label ?anatomicalentityName .\n",
      "\t?anatomicalentity <http://www.w3.org/2000/01/rdf-schema#label> ?anatomicalentity_label.\n",
      "\t?anatomicalentity <http://purl.org/dc/terms/description> ?anatomicalentity_description.\n",
      "\t?gene <http://www.w3.org/2000/01/rdf-schema#seeAlso> ?gene_seealso.\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "def rename_vars_meaningfully(query):\n",
    "    # here, idea is to test whether meaningful var names make a difference\n",
    "    # e.g. if we refer to a Gene as ?gene as opposed to ?x\n",
    "    # assumption: vars have classes defined in the original query text\n",
    "    [prefixes, vars_list, triples, optionals, filters] = get_core_triples(query)\n",
    "\n",
    "    replacements = {}\n",
    "    \n",
    "    # get all class declarations and match var name with expected name\n",
    "    for triple in triples:\n",
    "        s_p_o = triple.replace(\".\", \"\").split(\" \")[:3]\n",
    "        if(s_p_o[1] == \"a\" or s_p_o[1] == \"rdf:type\"): # here, consider multiple cases, e.g. rdf:type etc\n",
    "            var_name = s_p_o[0]\n",
    "            className = extract_name_from_URI(s_p_o[2])\n",
    "            if(var_name[1:].lower != className.lower()):\n",
    "                # replace var name with className everywhere\n",
    "                replacements[var_name] = \"?\" + className.lower()\n",
    "\n",
    "    # reconstruct query and replace all occurences of old var names with the new ones\n",
    "    reconstructed = construct_query(prefixes, vars_list, triples, optionals, filters)\n",
    "\n",
    "    for varName in replacements.keys():\n",
    "        reconstructed = reconstructed.replace(varName, replacements[varName])\n",
    "    \n",
    "    return reconstructed\n",
    "        \n",
    "\n",
    "test_query = \"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "SELECT DISTINCT ?anatomicalentity ?anatomicalentityName ?efo_343_4343 ?gene ?gene_seealso WHERE { \n",
    "\t?gene a orth:Gene .\n",
    "\t?gene genex:isExpressedIn ?anatomicalentity .\n",
    "\t?gene rdfs:label \"APOC1\" .\n",
    "\t?anatomicalentity a genex:AnatomicalEntity .\n",
    "\t?anatomicalentity rdfs:label ?anatomicalentityName .\n",
    "\t?anatomicalentity <http://www.w3.org/2000/01/rdf-schema#label> ?anatomicalentity_label.\n",
    "\t?anatomicalentity <http://purl.org/dc/terms/description> ?anatomicalentity_description.\n",
    "\t?gene <http://www.w3.org/2000/01/rdf-schema#seeAlso> ?gene_seealso.\n",
    "}\n",
    "\"\"\"\n",
    "vars_renamed_query = rename_vars_meaningfully(test_query)\n",
    "print(vars_renamed_query)\n",
    "\n",
    "#extract_name_from_URI(\"orth:Gene\")\n",
    "#extract_name_from_URI(\"http://purl.org/net/orth#Gene\")\n",
    "#extract_name_from_URI(\"http://purl.org/net/orth/Gene\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b3913a15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "SELECT DISTINCT ?x4Name ?x3 ?x1 ?x2 ?x0 ?x7 ?x6 ?x4 WHERE { \n",
      "\t?x0 a orth:Gene .\n",
      "\t?x1 a genex:Expression .\n",
      "\t?x1 genex:hasExpressionCondition ?x2 .\n",
      "\t?x1 genex:hasExpressionLevel ?x3 .\n",
      "\t?x1 genex:hasSequenceUnit ?x0 .\n",
      "\t?x0 dcterms:identifier \"118230125\" .\n",
      "\t?x2 genex:hasAnatomicalEntity ?x4 .\n",
      "\t?x4 a genex:AnatomicalEntity.\n",
      "\t?x4 rdfs:label ?x4Name .\n",
      "\t?x2 a genex:ExpressionCondition.\n",
      "\t?x2 genex:hasDevelopmentalStage ?x6 .\n",
      "\t?x2 genex:hasSex \"any\".\n",
      "\t?x2 genex:hasStrain ?x7 .\n",
      "\t?x7 rdfs:label \"wild-type\" .\n",
      "\tFILTER( ?x4 !=  obo:GO_0005575 )\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "def rename_vars_randomly(query):\n",
    "    # here, idea is to test whether meaningful var names make a difference\n",
    "    # we create a function to replace all var names with random ones\n",
    "    \n",
    "    [prefixes, vars_list, triples, optionals, filters] = get_core_triples(query)\n",
    "    \n",
    "    replacements = {}\n",
    "    \n",
    "    var_index = 0\n",
    "    \n",
    "    # get all class declarations and match var name with expected name\n",
    "    for triple in triples:\n",
    "        s_p_o = triple.replace(\".\", \"\").split(\" \")[:3]\n",
    "        if(s_p_o[0].startswith(\"?\")):\n",
    "            var_name = s_p_o[0] \n",
    "            if(replacements.get(var_name) is None):\n",
    "                replacements[var_name] = \"?x\" + str(var_index)\n",
    "                var_index += 1\n",
    "        if(s_p_o[1].startswith(\"?\")):\n",
    "            var_name = s_p_o[1]\n",
    "            if(replacements.get(var_name) is None):\n",
    "                replacements[var_name] = \"?x\" + str(var_index)\n",
    "                var_index += 1\n",
    "        if(s_p_o[2].startswith(\"?\")):\n",
    "            var_name = s_p_o[2]\n",
    "            if(replacements.get(var_name) is None):\n",
    "                replacements[var_name] = \"?x\" + str(var_index)\n",
    "                var_index += 1\n",
    "\n",
    "    # reconstruct query and replace all occurences of old var names with the new ones\n",
    "    reconstructed = construct_query(prefixes, vars_list, triples, optionals, filters)\n",
    "\n",
    "    for varName in replacements.keys():\n",
    "        reconstructed = reconstructed.replace(varName, replacements[varName])\n",
    "    \n",
    "    return reconstructed\n",
    "\n",
    "random_query = rename_vars_randomly(query)\n",
    "print(random_query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f0169d14",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Task I: Augment queries with 1 more hop (e.g. a label/comment/description triple pattern)\n",
    "\n",
    "# 1. get fragment from each Class URI in the list above, e.g. \"gene\"\n",
    "# Simplifying assumption: every fragment is unique (e.g. there are no 2 different Gene classes \n",
    "# with distinct namespaces in the same query) - otherwise we would need to parse the prefixes in the query\n",
    "# 2. look for triple with the fragment in the query text. if exists, get variable name. e.g. \"?gene a orth:Gene.\"\n",
    "# 3. attach triple pattern with the extra property to the corresponding variable\n",
    "# 4. Generate corresponding NL question deterministically \n",
    "# future work: improve using e.g. ChatGPT? OpenAI API?\n",
    "\n",
    "def augment_query_1_hop(query, question):\n",
    "    # first rename all var names to meaningful ones to make it easy to check if we can add more props\n",
    "    renamed_query = rename_vars_meaningfully(query)\n",
    "    \n",
    "    # parse triples from query\n",
    "    [prefixes, vars_list, triples, optionals, filters] = get_core_triples(renamed_query)\n",
    "    \n",
    "    augment_with_triple = None\n",
    "    \n",
    "    augmented_questions = []\n",
    "    augmented_queries = []\n",
    "    \n",
    "    already_added = set()\n",
    "    \n",
    "    # iterate through type declarations in query to check whether new datatype properties can be added\n",
    "    for triple in triples:\n",
    "        s_p_o = triple.replace(\".\", \"\").split(\" \")[:3]\n",
    "        if(s_p_o[1] == \"a\" or s_p_o[1] == \"rdf:type\"): # here, consider multiple cases, e.g. rdf:type etc\n",
    "            \n",
    "            var_name = s_p_o[0] \n",
    "            classNameQuery = extract_name_from_URI(s_p_o[2])\n",
    "            \n",
    "            # iterate through list of known properties for this className and see if triple can be added\n",
    "            # put the varName to be the label of the property wihtout spaces\n",
    "            for className in classNamesProps.keys():\n",
    "                if(classNameQuery.lower() == className.lower()):\n",
    "                    propNamesLabels = classNamesProps[className][\"propNamesLabels\"]\n",
    "                    # iterate through all properties attached to this class\n",
    "                    # all of these can be used to generate 1 extra variant of the query\n",
    "                    for propName in propNamesLabels.keys():\n",
    "                        propNameLabel = classNamesProps[className][\"propNamesLabels\"][propName]\n",
    "                        propNameLabel = propNameLabel.replace(\" \", \"\").replace(\"-\", \"_\").lower()\n",
    "                        propLabelVarName = className.lower() + \"_\" + propNameLabel\n",
    "                        to_add_triple = var_name + \" \" + \"<\" + propName + \">\" + \" \" + \"?\" + propLabelVarName + \".\"\n",
    "                        if(to_add_triple not in set(triples) and to_add_triple not in already_added):\n",
    "                            classNameText = list(classNamesProps[className][\"classLabel\"])[0].lower()\n",
    "                            propNameText = classNamesProps[className][\"propNamesLabels\"][propName].lower()\n",
    "                            if(question.find(\"Provide\") != -1):\n",
    "                                augmented_question = question[:-1] + \" and the \" + classNameText + \" \" + propNameText + \".\"\n",
    "                            else:    \n",
    "                                augmented_question = question + \" Provide also the \" + classNameText + \" \" + propNameText + \".\"\n",
    "                            augmented_questions += [augmented_question]\n",
    "                            augmented_queries += [construct_query(prefixes, list(vars_list)+[propLabelVarName], triples+[to_add_triple], optionals, filters)]\n",
    "                            already_added.add(to_add_triple)\n",
    "                            \n",
    "    if(len(augmented_queries) == 0):\n",
    "        # we added all properties possible\n",
    "        return [None, None]\n",
    "    \n",
    "    return [list(augmented_questions), list(augmented_queries)]\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d50fe110",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Before: \n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "\n",
      "SELECT DISTINCT ?anat ?anatName ?stageIRI ?score  {\n",
      "\t?seq a orth:Gene .\n",
      "\t\t?expression a genex:Expression .\n",
      "\t\t?expression genex:hasExpressionCondition ?condition .\n",
      "\t\t?expression genex:hasExpressionLevel ?score .\n",
      "\t?expression genex:hasSequenceUnit ?seq .\n",
      "\t\t?seq dcterms:identifier \"118230125\" .\n",
      "\t?condition genex:hasAnatomicalEntity ?anat .\n",
      "    ?anat a genex:AnatomicalEntity.\n",
      "\t\t?anat rdfs:label ?anatName .\n",
      "        ?condition a genex:ExpressionCondition.\n",
      "\t\t?condition genex:hasDevelopmentalStage ?stageIRI .\n",
      "\t?condition genex:hasSex \"any\".\n",
      "\t?condition genex:hasStrain ?strain .\n",
      "\t\t?strain rdfs:label \"wild-type\" .\n",
      "FILTER (?anat !=  obo:GO_0005575)\n",
      "} ORDER BY DESC(?score)\n",
      "\n",
      "Augmented Question: \n",
      "What are the anatomical entities where the eel gene apoc1 is expressed along with its expression score independently of the strain, sex, and cell type? Provide also the gene label.\n",
      "What are the anatomical entities where the eel gene apoc1 is expressed along with its expression score independently of the strain, sex, and cell type? Provide also the gene identifier.\n",
      "What are the anatomical entities where the eel gene apoc1 is expressed along with its expression score independently of the strain, sex, and cell type? Provide also the gene description.\n",
      "What are the anatomical entities where the eel gene apoc1 is expressed along with its expression score independently of the strain, sex, and cell type? Provide also the gene see also.\n",
      "What are the anatomical entities where the eel gene apoc1 is expressed along with its expression score independently of the strain, sex, and cell type? Provide also the gene expression has false discovery rate p-value.\n",
      "What are the anatomical entities where the eel gene apoc1 is expressed along with its expression score independently of the strain, sex, and cell type? Provide also the gene expression has expression level score.\n",
      "What are the anatomical entities where the eel gene apoc1 is expressed along with its expression score independently of the strain, sex, and cell type? Provide also the anatomical entity description.\n",
      "What are the anatomical entities where the eel gene apoc1 is expressed along with its expression score independently of the strain, sex, and cell type? Provide also the anatomical entity label.\n",
      "Augmented: 8\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "SELECT DISTINCT ?expressioncondition ?anatomicalentity ?score ?expression ?anatomicalentityName ?strain ?stageIRI ?gene ?gene_label WHERE { \n",
      "\t?gene a orth:Gene .\n",
      "\t?expression a genex:Expression .\n",
      "\t?expression genex:hasExpressionCondition ?expressioncondition .\n",
      "\t?expression genex:hasExpressionLevel ?score .\n",
      "\t?expression genex:hasSequenceUnit ?gene .\n",
      "\t?gene dcterms:identifier \"118230125\" .\n",
      "\t?expressioncondition genex:hasAnatomicalEntity ?anatomicalentity .\n",
      "\t?anatomicalentity a genex:AnatomicalEntity.\n",
      "\t?anatomicalentity rdfs:label ?anatomicalentityName .\n",
      "\t?expressioncondition a genex:ExpressionCondition.\n",
      "\t?expressioncondition genex:hasDevelopmentalStage ?stageIRI .\n",
      "\t?expressioncondition genex:hasSex \"any\".\n",
      "\t?expressioncondition genex:hasStrain ?strain .\n",
      "\t?strain rdfs:label \"wild-type\" .\n",
      "\t?gene <http://www.w3.org/2000/01/rdf-schema#label> ?gene_label.\n",
      "\tFILTER(  ?anatomicalentity !=  obo:GO_0005575  )\n",
      "}\n",
      "************\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "SELECT DISTINCT ?expressioncondition ?anatomicalentity ?score ?expression ?anatomicalentityName ?strain ?stageIRI ?gene ?gene_identifier WHERE { \n",
      "\t?gene a orth:Gene .\n",
      "\t?expression a genex:Expression .\n",
      "\t?expression genex:hasExpressionCondition ?expressioncondition .\n",
      "\t?expression genex:hasExpressionLevel ?score .\n",
      "\t?expression genex:hasSequenceUnit ?gene .\n",
      "\t?gene dcterms:identifier \"118230125\" .\n",
      "\t?expressioncondition genex:hasAnatomicalEntity ?anatomicalentity .\n",
      "\t?anatomicalentity a genex:AnatomicalEntity.\n",
      "\t?anatomicalentity rdfs:label ?anatomicalentityName .\n",
      "\t?expressioncondition a genex:ExpressionCondition.\n",
      "\t?expressioncondition genex:hasDevelopmentalStage ?stageIRI .\n",
      "\t?expressioncondition genex:hasSex \"any\".\n",
      "\t?expressioncondition genex:hasStrain ?strain .\n",
      "\t?strain rdfs:label \"wild-type\" .\n",
      "\t?gene <http://purl.org/dc/terms/identifier> ?gene_identifier.\n",
      "\tFILTER(  ?anatomicalentity !=  obo:GO_0005575  )\n",
      "}\n",
      "************\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "SELECT DISTINCT ?expressioncondition ?anatomicalentity ?score ?expression ?anatomicalentityName ?strain ?stageIRI ?gene ?gene_description WHERE { \n",
      "\t?gene a orth:Gene .\n",
      "\t?expression a genex:Expression .\n",
      "\t?expression genex:hasExpressionCondition ?expressioncondition .\n",
      "\t?expression genex:hasExpressionLevel ?score .\n",
      "\t?expression genex:hasSequenceUnit ?gene .\n",
      "\t?gene dcterms:identifier \"118230125\" .\n",
      "\t?expressioncondition genex:hasAnatomicalEntity ?anatomicalentity .\n",
      "\t?anatomicalentity a genex:AnatomicalEntity.\n",
      "\t?anatomicalentity rdfs:label ?anatomicalentityName .\n",
      "\t?expressioncondition a genex:ExpressionCondition.\n",
      "\t?expressioncondition genex:hasDevelopmentalStage ?stageIRI .\n",
      "\t?expressioncondition genex:hasSex \"any\".\n",
      "\t?expressioncondition genex:hasStrain ?strain .\n",
      "\t?strain rdfs:label \"wild-type\" .\n",
      "\t?gene <http://purl.org/dc/terms/description> ?gene_description.\n",
      "\tFILTER(  ?anatomicalentity !=  obo:GO_0005575  )\n",
      "}\n",
      "************\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "SELECT DISTINCT ?expressioncondition ?anatomicalentity ?score ?expression ?anatomicalentityName ?strain ?stageIRI ?gene ?gene_seealso WHERE { \n",
      "\t?gene a orth:Gene .\n",
      "\t?expression a genex:Expression .\n",
      "\t?expression genex:hasExpressionCondition ?expressioncondition .\n",
      "\t?expression genex:hasExpressionLevel ?score .\n",
      "\t?expression genex:hasSequenceUnit ?gene .\n",
      "\t?gene dcterms:identifier \"118230125\" .\n",
      "\t?expressioncondition genex:hasAnatomicalEntity ?anatomicalentity .\n",
      "\t?anatomicalentity a genex:AnatomicalEntity.\n",
      "\t?anatomicalentity rdfs:label ?anatomicalentityName .\n",
      "\t?expressioncondition a genex:ExpressionCondition.\n",
      "\t?expressioncondition genex:hasDevelopmentalStage ?stageIRI .\n",
      "\t?expressioncondition genex:hasSex \"any\".\n",
      "\t?expressioncondition genex:hasStrain ?strain .\n",
      "\t?strain rdfs:label \"wild-type\" .\n",
      "\t?gene <http://www.w3.org/2000/01/rdf-schema#seeAlso> ?gene_seealso.\n",
      "\tFILTER(  ?anatomicalentity !=  obo:GO_0005575  )\n",
      "}\n",
      "************\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "SELECT DISTINCT ?expressioncondition ?anatomicalentity ?score ?expression ?anatomicalentityName ?strain ?stageIRI ?gene ?expression_hasfalsediscoveryratep_value WHERE { \n",
      "\t?gene a orth:Gene .\n",
      "\t?expression a genex:Expression .\n",
      "\t?expression genex:hasExpressionCondition ?expressioncondition .\n",
      "\t?expression genex:hasExpressionLevel ?score .\n",
      "\t?expression genex:hasSequenceUnit ?gene .\n",
      "\t?gene dcterms:identifier \"118230125\" .\n",
      "\t?expressioncondition genex:hasAnatomicalEntity ?anatomicalentity .\n",
      "\t?anatomicalentity a genex:AnatomicalEntity.\n",
      "\t?anatomicalentity rdfs:label ?anatomicalentityName .\n",
      "\t?expressioncondition a genex:ExpressionCondition.\n",
      "\t?expressioncondition genex:hasDevelopmentalStage ?stageIRI .\n",
      "\t?expressioncondition genex:hasSex \"any\".\n",
      "\t?expressioncondition genex:hasStrain ?strain .\n",
      "\t?strain rdfs:label \"wild-type\" .\n",
      "\t?expression <http://purl.org/genex#hasFDRpvalue> ?expression_hasfalsediscoveryratep_value.\n",
      "\tFILTER(  ?anatomicalentity !=  obo:GO_0005575  )\n",
      "}\n",
      "************\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "SELECT DISTINCT ?expressioncondition ?anatomicalentity ?score ?expression ?anatomicalentityName ?strain ?stageIRI ?gene ?expression_hasexpressionlevelscore WHERE { \n",
      "\t?gene a orth:Gene .\n",
      "\t?expression a genex:Expression .\n",
      "\t?expression genex:hasExpressionCondition ?expressioncondition .\n",
      "\t?expression genex:hasExpressionLevel ?score .\n",
      "\t?expression genex:hasSequenceUnit ?gene .\n",
      "\t?gene dcterms:identifier \"118230125\" .\n",
      "\t?expressioncondition genex:hasAnatomicalEntity ?anatomicalentity .\n",
      "\t?anatomicalentity a genex:AnatomicalEntity.\n",
      "\t?anatomicalentity rdfs:label ?anatomicalentityName .\n",
      "\t?expressioncondition a genex:ExpressionCondition.\n",
      "\t?expressioncondition genex:hasDevelopmentalStage ?stageIRI .\n",
      "\t?expressioncondition genex:hasSex \"any\".\n",
      "\t?expressioncondition genex:hasStrain ?strain .\n",
      "\t?strain rdfs:label \"wild-type\" .\n",
      "\t?expression <http://purl.org/genex#hasExpressionLevel> ?expression_hasexpressionlevelscore.\n",
      "\tFILTER(  ?anatomicalentity !=  obo:GO_0005575  )\n",
      "}\n",
      "************\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "SELECT DISTINCT ?expressioncondition ?anatomicalentity ?score ?expression ?anatomicalentityName ?strain ?stageIRI ?gene ?anatomicalentity_description WHERE { \n",
      "\t?gene a orth:Gene .\n",
      "\t?expression a genex:Expression .\n",
      "\t?expression genex:hasExpressionCondition ?expressioncondition .\n",
      "\t?expression genex:hasExpressionLevel ?score .\n",
      "\t?expression genex:hasSequenceUnit ?gene .\n",
      "\t?gene dcterms:identifier \"118230125\" .\n",
      "\t?expressioncondition genex:hasAnatomicalEntity ?anatomicalentity .\n",
      "\t?anatomicalentity a genex:AnatomicalEntity.\n",
      "\t?anatomicalentity rdfs:label ?anatomicalentityName .\n",
      "\t?expressioncondition a genex:ExpressionCondition.\n",
      "\t?expressioncondition genex:hasDevelopmentalStage ?stageIRI .\n",
      "\t?expressioncondition genex:hasSex \"any\".\n",
      "\t?expressioncondition genex:hasStrain ?strain .\n",
      "\t?strain rdfs:label \"wild-type\" .\n",
      "\t?anatomicalentity <http://purl.org/dc/terms/description> ?anatomicalentity_description.\n",
      "\tFILTER(  ?anatomicalentity !=  obo:GO_0005575  )\n",
      "}\n",
      "************\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "SELECT DISTINCT ?expressioncondition ?anatomicalentity ?score ?expression ?anatomicalentityName ?strain ?stageIRI ?gene ?anatomicalentity_label WHERE { \n",
      "\t?gene a orth:Gene .\n",
      "\t?expression a genex:Expression .\n",
      "\t?expression genex:hasExpressionCondition ?expressioncondition .\n",
      "\t?expression genex:hasExpressionLevel ?score .\n",
      "\t?expression genex:hasSequenceUnit ?gene .\n",
      "\t?gene dcterms:identifier \"118230125\" .\n",
      "\t?expressioncondition genex:hasAnatomicalEntity ?anatomicalentity .\n",
      "\t?anatomicalentity a genex:AnatomicalEntity.\n",
      "\t?anatomicalentity rdfs:label ?anatomicalentityName .\n",
      "\t?expressioncondition a genex:ExpressionCondition.\n",
      "\t?expressioncondition genex:hasDevelopmentalStage ?stageIRI .\n",
      "\t?expressioncondition genex:hasSex \"any\".\n",
      "\t?expressioncondition genex:hasStrain ?strain .\n",
      "\t?strain rdfs:label \"wild-type\" .\n",
      "\t?anatomicalentity <http://www.w3.org/2000/01/rdf-schema#label> ?anatomicalentity_label.\n",
      "\tFILTER(  ?anatomicalentity !=  obo:GO_0005575  )\n",
      "}\n",
      "************\n"
     ]
    }
   ],
   "source": [
    "print(\"Before: \" + query)\n",
    "[aug_questions, augmented_queries] = augment_query_1_hop(query, nlqs[12])\n",
    "\n",
    "print(\"Augmented Question: \")\n",
    "for question in aug_questions:\n",
    "    print(question)\n",
    "    \n",
    "print(\"Augmented: \" + str(len(augmented_queries)))\n",
    "for query in augmented_queries:\n",
    "    print(query)\n",
    "    print(\"************\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7231f501",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                         propertyURI                label\n",
      "0  http://bioontology.org/ontologies/biositemap.o...  version information\n",
      "1        http://purl.obolibrary.org/obo/CDAO_0000190           belongs_to\n",
      "2          http://omabrowser.org/ontology/oma#domain               domain\n",
      "domain\n"
     ]
    }
   ],
   "source": [
    "# Task II: Label existing query with comments (e.g. describing property or class names)\n",
    "\n",
    "# first we need the object properties labels\n",
    "df_labels_props = pd.read_csv ('object_properties_labels.csv')\n",
    "df_labels_props = df_labels_props.drop_duplicates()\n",
    "print(df_labels_props.head(3))\n",
    "\n",
    "print(df_labels_props[df_labels_props['propertyURI']== \"http://omabrowser.org/ontology/oma#domain\"][\"label\"].tolist()[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "608ca8fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "SELECT DISTINCT ?expressioncondition ?anatomicalentity ?expression ?score ?anatomicalentityName ?strain ?stageIRI ?anatomicalentity_label ?gene WHERE { \n",
      "\t?gene a orth:Gene .\n",
      "\t?expression a genex:Expression .\n",
      "\t?expression genex:hasExpressionCondition ?expressioncondition . # has an expression condition\n",
      "\t?expression genex:hasExpressionLevel ?score .\n",
      "\t?expression genex:hasSequenceUnit ?gene . # has sequence unit\n",
      "\t?gene dcterms:identifier \"118230125\" .\n",
      "\t?expressioncondition genex:hasAnatomicalEntity ?anatomicalentity . # has anatomical entity\n",
      "\t?anatomicalentity a genex:AnatomicalEntity.\n",
      "\t?anatomicalentity rdfs:label ?anatomicalentityName .\n",
      "\t?expressioncondition a genex:ExpressionCondition.\n",
      "\t?expressioncondition genex:hasDevelopmentalStage ?stageIRI . # has a developmental stage\n",
      "\t?expressioncondition genex:hasSex \"any\".\n",
      "\t?expressioncondition genex:hasStrain ?strain . # has strain\n",
      "\t?strain rdfs:label \"wild-type\" .\n",
      "\t?anatomicalentity <http://www.w3.org/2000/01/rdf-schema#label> ?anatomicalentity_label.\n",
      "\tFILTER(   ?anatomicalentity !=  obo:GO_0005575   )\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "def augment_query_with_comments(query):\n",
    "    [prefixes, vars_list, triples, optionals, filters] = get_core_triples(query)\n",
    "    \n",
    "    final_triples = [] \n",
    "    for triple in triples:\n",
    "        # get the property name and add the property label as an inline comment to the triple\n",
    "        s_p_o = triple.replace(\".\", \"\").split(\" \")[:3]\n",
    "      \n",
    "        if(not s_p_o[1].startswith(\"?\")):\n",
    "            propUri = s_p_o[1]\n",
    "            propName = extract_name_from_URI(propUri).lower()\n",
    "            for propertyURI in df_labels_props['propertyURI'].tolist():\n",
    "                if(extract_name_from_URI(propertyURI).lower() == propName):\n",
    "                    triple += \" # \" + df_labels_props[df_labels_props['propertyURI'] == propertyURI]['label'].tolist()[0]\n",
    "        final_triples += [triple]\n",
    "                    \n",
    "    return construct_query(prefixes, vars_list, final_triples, optionals, filters)\n",
    "\n",
    "res = augment_query_with_comments(query)\n",
    "print(res)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3424106a",
   "metadata": {},
   "source": [
    "# Next, generate 3 sets of queries:\n",
    " 1. random variable names\n",
    " 2. meaningful variable names\n",
    " 3. meaningful variable names + inline comments for property names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "80162de9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# iterate through existing queries, progressively augment with 1 hop\n",
    "# create 4 folders:\n",
    "# 0. original queries (with 1-hop augments)\n",
    "# 1. random_vars\n",
    "# 2. meaningful_vars\n",
    "# 3. meaningful_vars + comments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "0c6d9170",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generated...100 variants\n",
      "Generated...200 variants\n",
      "Generated...300 variants\n",
      "Generated...400 variants\n",
      "Generated...500 variants\n",
      "Generated...600 variants\n"
     ]
    }
   ],
   "source": [
    "printed = False\n",
    "\n",
    "index_main_query = 0\n",
    "\n",
    "generated_variants = 0\n",
    "\n",
    "for queries in qs: # iterate through queries, e.g. from 1 to 15\n",
    "    for query in queries: # iterate through variants, e.g. 8a 8b etc\n",
    "     if(query):\n",
    "        index_variant = 0\n",
    "        \n",
    "        index_augment_variant = 0\n",
    "        \n",
    "        # write original question to file\n",
    "        \n",
    "        current_question = nlqs[index_main_query]\n",
    "        with open(\"dataset/questions/query_\"+ str(index_main_query) + \"_\" + str(index_variant) + \"_\" + str(index_augment_variant) + \".txt\", \"w\") as f:\n",
    "            f.write(current_question)\n",
    "        \n",
    "        # write original query to file\n",
    "        with open(\"dataset/original_augmented/query_\"+ str(index_main_query) + \"_\" + str(index_variant) + \"_\" + str(index_augment_variant) + \".rq\", \"w\") as f:\n",
    "            f.write(\"# \" + current_question + \"\\n\")\n",
    "            f.write(query)\n",
    "        \n",
    "        with open(\"dataset/random_vars/query_\"+ str(index_main_query) + \"_\" + str(index_variant) + \"_\" + str(index_augment_variant) + \".rq\", \"w\") as f:\n",
    "            f.write(\"# \" + current_question + \"\\n\")\n",
    "            f.write(rename_vars_randomly(query))\n",
    "        \n",
    "        with open(\"dataset/meaningful_vars/query_\"+ str(index_main_query) + \"_\" + str(index_variant) + \"_\" + str(index_augment_variant) + \".rq\", \"w\") as f:\n",
    "            f.write(\"# \" + current_question + \"\\n\")\n",
    "            f.write(rename_vars_meaningfully(query))\n",
    "            \n",
    "        with open(\"dataset/meaningful_vars_comments/query_\"+ str(index_main_query) + \"_\" + str(index_variant) + \"_\" + str(index_augment_variant) + \".rq\", \"w\") as f:\n",
    "            f.write(\"# \" + current_question + \"\\n\")\n",
    "            f.write(augment_query_with_comments(rename_vars_meaningfully(query)))\n",
    "        \n",
    "        \n",
    "        # augment each query with 1 hop iteratively until no more additions are possible\n",
    "        [augmented_questions, augmented_queries] = augment_query_1_hop(query, nlqs[index_main_query])\n",
    "            \n",
    "        while(augmented_queries != None): # iterate through augmentations\n",
    "        \n",
    "            for augmented_query in augmented_queries:\n",
    "                index_augment_variant += 1\n",
    "                generated_variants += 1\n",
    "                if(generated_variants % 100 == 0):\n",
    "                    print(\"Generated...{} variants\".format(generated_variants))\n",
    "                    \n",
    "                current_question = augmented_questions[augmented_queries.index(augmented_query)]\n",
    "                    \n",
    "                with open(\"dataset/questions/query_\"+ str(index_main_query) + \"_\" + str(index_variant) + \"_\" + str(index_augment_variant) + \".txt\", \"w\") as f:\n",
    "                    f.write(current_question)\n",
    "                    \n",
    "                with open(\"dataset/original_augmented/query_\"+ str(index_main_query) + \"_\" + str(index_variant) + \"_\" + str(index_augment_variant) + \".rq\", \"w\") as f:\n",
    "                    f.write(\"# \" + current_question + \"\\n\")\n",
    "                    f.write(augmented_query)\n",
    "        \n",
    "                with open(\"dataset/random_vars/query_\"+ str(index_main_query) + \"_\" + str(index_variant) + \"_\" + str(index_augment_variant) + \".rq\", \"w\") as f:\n",
    "                    f.write(\"# \" + current_question + \"\\n\")\n",
    "                    f.write(rename_vars_randomly(augmented_query))\n",
    "                \n",
    "                with open(\"dataset/meaningful_vars/query_\"+ str(index_main_query) + \"_\" + str(index_variant) + \"_\" + str(index_augment_variant) + \".rq\", \"w\") as f:\n",
    "                    f.write(\"# \" + current_question + \"\\n\")\n",
    "                    f.write(rename_vars_meaningfully(augmented_query))\n",
    "                \n",
    "                with open(\"dataset/meaningful_vars_comments/query_\"+ str(index_main_query) + \"_\" + str(index_variant) + \"_\" + str(index_augment_variant) + \".rq\", \"w\") as f:\n",
    "                    f.write(\"# \" + current_question + \"\\n\")\n",
    "                    f.write(augment_query_with_comments(rename_vars_meaningfully(augmented_query)))\n",
    "                    \n",
    "            [augmented_questions, augmented_queries] = augment_query_1_hop(augmented_query, augmented_questions[augmented_queries.index(augmented_query)])\n",
    "\n",
    "        index_variant += 1\n",
    "            \n",
    "    index_main_query += 1\n",
    "    if(index_main_query == len(nlqs) - 1):\n",
    "        break\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "55b46a73",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generated...100 variants\n",
      "Generated...200 variants\n",
      "Generated...300 variants\n",
      "Generated...400 variants\n",
      "Generated...500 variants\n",
      "Generated...600 variants\n"
     ]
    }
   ],
   "source": [
    "printed = False\n",
    "\n",
    "index_main_query = 0\n",
    "\n",
    "generated_variants = 0\n",
    "\n",
    "for queries in qs: # iterate through queries, e.g. from 1 to 15\n",
    "    for query in queries: # iterate through variants, e.g. 8a 8b etc\n",
    "     if(query):\n",
    "        index_variant = 0\n",
    "        \n",
    "        index_augment_variant = 0\n",
    "        \n",
    "        # write original question to file\n",
    "        \n",
    "        current_question = nlqs[index_main_query]\n",
    "        \n",
    "        with open(\"dataset/new_original_augmented/query_\"+ str(index_main_query) + \"_\" + str(index_variant) + \"_\" + str(index_augment_variant) + \".rq\", \"w\") as f:\n",
    "                    f.write(\"# \" + current_question + \"\\n\")\n",
    "                    f.write(query)\n",
    "            \n",
    "        with open(\"dataset/original_with_comments/query_\"+ str(index_main_query) + \"_\" + str(index_variant) + \"_\" + str(index_augment_variant) + \".rq\", \"w\") as f:\n",
    "            f.write(\"# \" + current_question + \"\\n\")\n",
    "            f.write(augment_query_with_comments(query))\n",
    "        \n",
    "        \n",
    "        # augment each query with 1 hop iteratively until no more additions are possible\n",
    "        [augmented_questions, augmented_queries] = augment_query_1_hop(query, nlqs[index_main_query])\n",
    "            \n",
    "        while(augmented_queries != None): # iterate through augmentations\n",
    "        \n",
    "            for augmented_query in augmented_queries:\n",
    "                index_augment_variant += 1\n",
    "                generated_variants += 1\n",
    "                if(generated_variants % 100 == 0):\n",
    "                    print(\"Generated...{} variants\".format(generated_variants))\n",
    "                \n",
    "                current_question = augmented_questions[augmented_queries.index(augmented_query)]\n",
    "                    \n",
    "                with open(\"dataset/new_original_augmented/query_\"+ str(index_main_query) + \"_\" + str(index_variant) + \"_\" + str(index_augment_variant) + \".rq\", \"w\") as f:\n",
    "                    f.write(\"# \" + current_question + \"\\n\")\n",
    "                    f.write(augmented_query)\n",
    "                \n",
    "                with open(\"dataset/original_with_comments/query_\"+ str(index_main_query) + \"_\" + str(index_variant) + \"_\" + str(index_augment_variant) + \".rq\", \"w\") as f:\n",
    "                    f.write(\"# \" + current_question + \"\\n\")\n",
    "                    f.write(augment_query_with_comments(augmented_query))\n",
    "                    \n",
    "            [augmented_questions, augmented_queries] = augment_query_1_hop(augmented_query, augmented_questions[augmented_queries.index(augmented_query)])\n",
    "\n",
    "        index_variant += 1\n",
    "            \n",
    "    index_main_query += 1\n",
    "    if(index_main_query == len(nlqs) - 1):\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0a30e2b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
